polaris/hub/client.py

```diff
@@ -288,17 +289,39 @@ def list_datasets(self, limit: int = 100, offset: int = 0) -> list[str]:
             A list of dataset names in the format `owner/dataset_name`.
         """
         with ProgressIndicator(
-            start_msg="Fetching artifacts...",
-            success_msg="Fetched artifacts.",
+            start_msg="Fetching datasets...",
+            success_msg="Fetched datasets.",
             error_msg="Failed to fetch datasets.",
         ):
-            response = self._base_request_to_hub(
-                url="/v1/dataset", method="GET", params={"limit": limit, "offset": offset}
-            )
-            response_data = response.json()
-            dataset_list = [bm["artifactId"] for bm in response_data["data"]]
-
-            return dataset_list
+            # Step 1: Fetch enough v2 datasets to cover the offset and limit
+            v2_json_response = self._base_request_to_hub(
+                url="/v2/dataset", method="GET", params={"limit": limit, "offset": offset}
+            ).json()
+            v2_data = v2_json_response["data"]
+            v2_datasets = [dataset["artifactId"] for dataset in v2_data]
+
+            # If v2 datasets satisfy the limit, return them
+            if len(v2_datasets) == limit:
+                return v2_datasets
+
+            # Step 2: Calculate the remaining limit and fetch v1 datasets
+            remaining_limit = max(0, limit - len(v2_datasets))
+
+            v1_datasets = []
+            if remaining_limit > 0:
```

Comment:
**Comment:** This will always be true, because we return early if `len(v2_datasets) == limit`